### PR TITLE
Fully remove hand-written listers and informers

### PR DESCRIPTION
**Pull request by ncdc**
Note: the first commit is from #41927. Adding do-not-merge for now as we'll want that to go in first, and then I'll rebase this on top.

Update statefulset controller to use a lister for PVCs instead of a client request. Also replace a unit test's dependency on legacylisters with the generated ones. cc @kargakis @kow3ns @foxish @kubernetes/sig-apps-pr-reviews 

Remove all references to pkg/controller/informers and pkg/client/legacylisters, and remove those packages.

@smarterclayton @deads2k this should be it!

cc @gmarek @wojtek-t @derekwaynecarr @kubernetes/sig-scalability-pr-reviews

### CODE REVIEW

**Comment by deads2k** [comment]
lgtm.  Victory is sweet. :)

**Comment by ncdc** [comment]
@k8s-bot gce etcd3 e2e test this #41893 #39821

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go [hunk 1]
```diff
@@ -349,12 +353,18 @@ func (frsc *ReplicaSetController) deliverLocalReplicaSet(obj interface{}, durati
...
-		return
+		glog.Errorf("Error splitting key for object %v: %v", obj, err)
 	}
-	if exists { // ignore replicasets exists only in local k8s
+	_, err = frsc.replicaSetLister.ReplicaSets(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// do nothing
+	case err != nil:
+		glog.Errorf("Couldn't get federation replicaset %v: %v", key, err)
```
Comment (kargakis) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: utilruntime.HandleError instead

File: federation/pkg/federation-controller/service/endpoint_helper.go [hunk 2]
```diff
@@ -81,29 +81,25 @@ func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache
...
-		}
-	} else {
+	endpoint, err := clusterCache.endpointStore.Endpoints(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
 		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
 		glog.Infof("Can not get endpoint %v for cluster %s from endpointStore", key, clusterName)
 		err = cc.processEndpointDeletion(cachedService, clusterName, serviceController)
+	case err != nil:
+		glog.Errorf("Did not successfully get %v from store: %v, will retry later", key, err)
```
Comment (kargakis) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: HandleError

File: federation/pkg/federation-controller/service/service_helper.go [hunk 3]
```diff
@@ -85,31 +84,26 @@ func (cc *clusterClientCache) syncService(key, clusterName string, clusterCache
...
-		}
-	} else {
+	service, err := clusterCache.serviceStore.Services(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
 		glog.Infof("Can not get service %v for cluster %s from serviceStore", key, clusterName)
 		needUpdate = cc.processServiceDeletion(cachedService, clusterName)
 		isDeletion = true
+	case err != nil:
+		glog.Errorf("Did not successfully get %v from store: %v, will retry later", key, err)
```
Comment (kargakis) [inline federation/pkg/federation-controller/service/service_helper.go]: HandleError

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go (hunk 1 above)
Comment (ncdc) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: Please don't review the first 2 commits in this PR right now

Comment (ncdc) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: This has already been covered in #41927

File: federation/pkg/federation-controller/service/service_helper.go (hunk 3 above)
Comment (kargakis) [inline federation/pkg/federation-controller/service/service_helper.go]: you've got it in a couple of other places too

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 4]
```diff
@@ -209,18 +215,17 @@ func (spc *realStatefulPodControl) recordClaimEvent(verb string, set *apps.State
...
-				errs = append(errs, fmt.Errorf("Failed to retrieve PVC %s: %s", claim.Name, err))
-				spc.recordClaimEvent("create", set, pod, &claim, err)
+		_, err := spc.pvcLister.PersistentVolumeClaims(claim.Namespace).Get(claim.Name)
+		switch {
+		case apierrors.IsNotFound(err):
+			_, err := spc.client.Core().PersistentVolumeClaims(claim.Namespace).Create(&claim)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("Failed to create PVC %s: %s", claim.Name, err))
 			}
+			spc.recordClaimEvent("create", set, pod, &claim, err)
```
Comment (kargakis) [inline pkg/controller/statefulset/stateful_pod_control.go]: By using a cache, this create error may happen more than once. I would not emit AlreadyExists errors.

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: Ok

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: Would there ever be a situation where the create failed with AlreadyExists and the preexisting PVC is somehow "wrong" in terms of its details?

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: @kargakis I just pushed a new commit to hopefully address this. PTAL

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 5]
```diff
@@ -218,11 +218,13 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
 		_, err := spc.pvcLister.PersistentVolumeClaims(claim.Namespace).Get(claim.Name)
 		switch {
 		case apierrors.IsNotFound(err):
-			_, err := spc.client.Core().PersistentVolumeClaims(claim.Namespace).Create(&claim)
+			created, err := spc.createPersistentVolumeClaim(set, &claim)
 			if err != nil {
 				errs = append(errs, fmt.Errorf("Failed to create PVC %s: %s", claim.Name, err))
 			}
-			spc.recordClaimEvent("create", set, pod, &claim, err)
+			if created || err != nil {
```
Comment (kargakis) [inline pkg/controller/statefulset/stateful_pod_control.go]: How about simply checking for the error type here and drop the client wrapper? #bareclientmovement
```go
if err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: @kargakis do you like this better?

```go
case apierrors.IsNotFound(err):
  _, err := spc.client.Core().PersistentVolumeClaims(claim.Namespace).Create(&claim)
  if err != nil && !apierrors.IsAlreadyExists(err){
    errs = append(errs, fmt.Errorf("Failed to create PVC %s: %s", claim.Name, err))
  }
  if err == nil || !apierrors.IsAlreadyExists(err) {
    spc.recordClaimEvent("create", set, pod, &claim, err)
  }
```

**Comment by wojtek-t** [comment]
wohoo! (I didn't look into PR, but the title of it is great)

File: pkg/controller/statefulset/stateful_pod_control.go (hunk 5 above)
Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: @kargakis I pushed an update with this. PTAL

Comment (kargakis) [inline pkg/controller/statefulset/stateful_pod_control.go]: @ncdc the AlreadyExists error should be returned as part of the error slice so we will force the statefulset to be requeued and pick up the PVC the next time, just the event shouldn't be emitted to users.

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: @kargakis the code in master right now always records an event after it attempts to create the PVC. This change only records if it succeeds, or if the error is something other than already-exists. Is that not what you want?

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: Oh I see what you're saying

Comment (kargakis) [inline pkg/controller/statefulset/stateful_pod_control.go]: statefulset changes lgtm

**Comment by ncdc** [comment]
Rebased and hopefully ready to go

**Comment by ncdc** [comment]
@deads2k @smarterclayton @liggitt @sttts PTAL

**Comment by smarterclayton** [comment]
Marking p2 to remove the final set of code and get rid of this forever.

**Pull request merged**